### PR TITLE
Revised back buttons in headers

### LINKED ISSUE #172: Add lightweight breadcrumbs to page header with back button
- Show one level of context in the page header (for example, the project name in Screendoor)
- Add space for "labels" (i.e. "Demo" / "Archived")
- Add space for states? ("Revision", "Draft", "Private", etc.)

### PR DESCRIPTION
**Pull request by jrubenoff**
Closes #172 

screen shot 2015-09-29 at 12 03 20 pm

Two links in the header:

screen shot 2015-09-29 at 12 06 36 pm

In Screendoor, the "root level" is always the project.

On most Screendoor pages, the root level and the Back button will link to the same thing. But the two separate links come in handy when you go multiple levels deep. For example, you would see this when editing a message preset:

screen shot 2015-09-29 at 12 09 16 pm

### CODE REVIEW

**Comment by ajb** [comment]
Made a couple commits, go ahead and review those :smile:

One issue I'd like to resolve before we merge:

The added padding for the `with_back_button` modifier means that the page header's height changes when there's a back arrow. This will look janky in an application context. Visit http://dvlcore.dev/regression/header_one for an example.

**Comment by jrubenoff** [comment]
Alright, fixed... looks good here. Not sure how it will look in Screendoor, but we can adjust as we go.

File: vendor/assets/stylesheets/dvl/components/page_header.scss [hunk 1]
```diff
@@ -55,6 +58,22 @@
 
   &.with_back_arrow {
     padding-left: $rhythm * 5;
+    h2 {
```
Comment (ajb) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: Can we un-nest this one level? I don't think it needs the `&.with_back_arrow` specificity, right?

Comment (ajb) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: (I did this)

**Comment by ajb** [comment]
:+1: Looks great.

Let's do the PR to implement in Screendoor before merging this commit. No reason to merge before we know if it's ready.

**Comment by ajb** [comment]
There's not really any rush here... but I just went and rebased this off of master.

Some issues:

img

img

img

**Comment by jrubenoff** [comment]
Thanks for catching! Just fixed all of 'em.

**Comment by jrubenoff** [comment]
On Friday, I added #207, as well as multi-level breadcrumbs, like we discussed in dobtco/screendoor-v2#2811.

Now to implement in Screendoor...

**Comment by ajb** [comment]
Looks great. I agree, this is ready for implementation in Screendoor. I'll take a stab at that?

A couple of quick points for feedback:

img

img

**Comment by jrubenoff** [comment]
Clarified the header styles.

Re: the carets, everything's kind of a mess on mobile already. Would rather address that for the whole style guide in a future sprint.

**Comment by ajb** [comment]
I was actually wondering about the difference between using `.button_uppercase` and a regular `.button`. (Should have been clearer, sorry!)

Anyway, this looks great. We'll need to rebase before merging, but let's wait on https://github.com/dobtco/screendoor-v2/issues/2811.

File: spec/dummy/app/views/home/headers.rb [hunk 2]
```diff
@@ -138,13 +135,28 @@ def main
 
     hr
 
-    docs 'Basic Page Header', %{
+    h3 'Basic Page Header'
+
+    docs 'Multi-Step Flows', %{
+      div(class: 'page_header') {
+        div(class: 'page_header_context'){
+          span '2015 Civic Innovation Conference', href: '#'
```
Comment (ajb) [inline spec/dummy/app/views/home/headers.rb]: Shouldn't these be `a` tags?

File: vendor/assets/stylesheets/dvl/components/page_header.scss [hunk 3]
```diff
@@ -33,35 +36,92 @@
   }
 
   > .button,
-  > form .button {
+  > form .button,
```
Comment (ajb) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: Out of curiosity, when do we ever have a `<form>` tag inside of a page headeR?

Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: Off the top of my head...I think on the "All projects" page?

File: spec/dummy/app/views/home/headers.rb (hunk 2 above)
Comment (jrubenoff) [inline spec/dummy/app/views/home/headers.rb]: No, because we don't want users clicking and discarding their work accidentally in the middle of a task.

Comment (ajb) [inline spec/dummy/app/views/home/headers.rb]: We should use a `beforeunload` warning for that. If we follow this logic, wouldn't we remove all links from the page?

(Also, the `href` attribute on a `span` tag is invalid...)

File: vendor/assets/stylesheets/dvl/components/page_header.scss (hunk 3 above)
Comment (ajb) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: Ah yeah, you're right. Do you think we need the direct child selector (`>`), or can we remove it to simplify things? (I guess I could test by adding an example to `forms.rb` or a new regression pgae.)

Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: I know this is here for a reason. I'm fine with removing it if we check every page in Screendoor...

File: spec/dummy/app/views/home/headers.rb (hunk 2 above)
Comment (jrubenoff) [inline spec/dummy/app/views/home/headers.rb]: You're right. But we need to specify that in the style guide. Also, a modal is probably nicer than a JS alert for this.

Overlooked the `href` attribute, thanks. Fixed

File: vendor/assets/stylesheets/dvl/components/page_header.scss (hunk 3 above)
Comment (ajb) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: If you know it's here for a reason that's fine... I'm just curious when there would be a button inside of a page header that we _don't_ want to style.

Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: Ah! The Rename button.

**Comment by ajb** [comment]
This is nitpicking, but it can't actually be a "confirmation modal". The `beforeunload` event is natively handled by the browser, so in the cases that we can intercept it w/ Turbolinks, we try to emulate it as closely as possible. Otherwise, the behavior would be different depending on _how_ the user was trying to leave the page.

File: vendor/assets/stylesheets/dvl/components/page_header.scss [hunk 4]
```diff
@@ -33,35 +36,92 @@
...
+    margin-bottom: $rhythm;
+    @media screen and (min-width: $lapWidth) {
+      margin: ($rhythm * 1.5) 0;
+    }
+  }
+
   // Modifiers
 
   &.with_back_arrow {
+    margin-top: $rhythm;
```
Comment (ajb) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: @jrubenoff: Can you explain why this is here? This leads to some unwanted behavior in Screendoor:

When changing from the "New project" to "Example projects" page, the page header jumps down, even though there's no context being added on the example projects page.

Screencast: http://take.ms/ry8LB

Comment (jrubenoff) [inline vendor/assets/stylesheets/dvl/components/page_header.scss]: Added a new modifier, `.with_no_context`, to address this.

**Comment by ajb** [comment]
Rebasing in https://github.com/dobtco/dvl-core/pull/226.

**Pull request closed without merging**